### PR TITLE
Improvment certificate api (limit)

### PR DESCRIPTION
**Pull request by ghost**
Second PR to https://github.com/ILIAS-eLearning/ILIAS/pull/2214

- Implement limit start and end (Table pagination) - this also need a second method to get the max count (without limit and order)

### CODE REVIEW

**Comment by mjansenDatabay** [comment]
Thx @sragfw , the `sort` feature was already approved (by me) in #2214 ;-). My concerns were only raised about the `Limit/Offset`.

**Comment by ghost** [comment]
@mjansenDatabay Sorry, yes, you're right, it was my problem (I had somehow misunderstand you ...)
So here PR https://github.com/ILIAS-eLearning/ILIAS/pull/2217 with the multiple sort changes
If you merged it, I will merge trunk in this and only the limit should left here

**Comment by mjansenDatabay** [comment]
@sragfw Currently I see one major issue with the pagination. The SQL statement might result in **multiple records**  for the **same unique certificate id**. This is a valid case if (for instance) a user certificate was issued for a test, which has **multiple references** in the ILIAS repository (e.g. in Course 1, Course 2, and Course 3).

So the question is: What is the expected behaviour of the consumer/client when using the pagination? Let's say the result of the SQL query returns 100 records, but only 57 unique certificates, with the other records being **different** `ref_ids` for the certificates.

**Comment by ghost** [comment]
If the consumer/client has set for instance 50 rows per page and is on page 1, he becomes the data rows 1-50

If he navigate to page 2, he becomes the data 51-100, etc. ...

The max count (Needed for calculate the pages of pagination) is the count of rows without limit and order

I dont' think, that the limit has influence of the unique of rows as you wrote (but I can wrong of course)

As alternative for the SQL limit, it could be to do it with PHP something like the follow before return the result (Like the old `ilTable2GUI` does, but new (https://github.com/ILIAS-eLearning/ILIAS/pull/2214#issuecomment-536954459), the data fetcher is always responsible to filter the limit/pagination of the data (For performance reasons the database is faster than PHP))

```php
...
        if ($filter->getLimitStart() && $filter->getLimitEnd()) {
            $result = array_slice($result, $filter->getLimitStart(), ($filter->getLimitEnd() - $filter->getLimitStart()));
        }

        return $result;
```

**Comment by mjansenDatabay** [comment]
Thanks for your reply @sragfw . Maybe we (together with @mstuder ?) could have a short Skype call? I'll need to (better) understand your usage scenario. How is the certificate information provided in your views? Will there be a list/table of all users (one row per certificate with multiple links to the references (if not trashed))? In this case the pagination **must** (IMO) be implemented in PHP, not in SQL.

**Comment by ghost** [comment]
@mjansenDatabay 

ok, now I understand your use scenario and you're right

I adjust that the limit/pagination is done with PHP (after grouping to certificate id and may multiple ref_id's)

**Comment by ghost** [comment]
I implement to filter also by ref_id and a possibility to not include deleted objects

**Comment by ghost** [comment]
And it should also possible to get all certificates - so I remove the params from filter constructor and allows empty user and ref ids array - I need also to change the WHERE builder (use array $wheres and not direct ad to $sql because all filter are now voluntary)

**Comment by ghost** [comment]
- Filter by obj ids (instead of ref ids)
- Rename `onlyActive` to `onlyCertActive`
- Make query with `UNION` (deleted) - also objects in trash are now deleted
- Not filter empty strings
- use `ilDBConstants`

File: Services/Certificate/classes/API/Filter/UserDataFilter.php [hunk 1]
```diff
@@ -218,19 +210,41 @@ public function withAdditionalUserIds(array $usrIds) : self
...
     }
 
     /**
-     * @return int
+     * @param int[] $objIds
+     *
+     * @return $this
      */
-    public function getObjectId() : ?int
+    public function withAdditionalobjIds(array $objIds) : self
```
Comment (mjansenDatabay) [inline Services/Certificate/classes/API/Filter/UserDataFilter.php]: `withAdditionalObjIds` (upper case O)

Comment (sragfw) [reply Services/Certificate/classes/API/Filter/UserDataFilter.php]: Already fixed

Comment (ghost) [reply Services/Certificate/classes/API/Filter/UserDataFilter.php]: Fixed

**Pull request merged**